usr.sbin/npf/npfctl/npf_build.c

```diff
@@ -688,13 +767,35 @@ npfctl_build_group_end(void)
 	 * - If the parent is NULL, then it is a global rule.
 	 * - The default rule must be the last, so it is inserted later.
 	 */
-	if (group == defgroup) {
+	if (group == defgroup || group == defgroup_l2) {
 		assert(parent == NULL);
 		return;
 	}
 	npf_rule_insert(npf_conf, parent, group);
 }
 
+/*
+ * this function is here to ensure that layer 2 rules are
+ * rightfully embedded in layer2 groups
+ * and vice versa. layer3 group => layer 3 rules
+ * does not allow setting layer 2 rules in layer 3 groups
+ */
+
+static uint32_t
+npf_rule_layer_compat(nl_rule_t *cg, uint32_t layer)
+{
+	const char *str = (layer & NPF_RULE_LAYER_2) ? "layer 2" : "layer 3";
```

Comment:
You are using 2 flags one for layer 2 and one for layer 3, then you test layer 2 here and in the code above you test layer 3 (with the else being the other one). Why do you need two flags? Are we planning to filter more layers in the future? And if we do, then if-then-else is not good enough.